### PR TITLE
[core] Clarify GPT output_processor callback contract

### PR DESCRIPTION
**Pull request by chengcuiping**
## What does this PR do?

This PR clarifies the return contract of `GPTModel`'s optional `output_processor` callback.

Specifically, it:

- generalizes the callback return annotation from `Tensor` to `Any`;
- documents that, for direct `GPTModel.forward()` calls, the processor result becomes the forward result;
- adds direct-forward unit coverage for a structured, non-Tensor callback result.

## Motivation

The runtime hook already returns the processor result directly, but its existing `Callable[..., Tensor]` annotation is narrower than the actual behavior.

Downstream validation exposed this mismatch using a structured model output. This PR keeps the contract objective-neutral and does not introduce framework- or RL-specific output types.

## Scope

This is a typing, documentation, and unit-test change. It does not change:

- `_postprocess()` runtime behavior;
- the default logits or loss path;
- MTP behavior;
- 1F1B scheduling;
- schedule-plan output conversion;
- `float16_to_fp32()` behavior.

## Testing

Focused callback and schedule-plan tests:

```text
CUDA_VISIBLE_DEVICES=0 pytest -q tests/unit_tests/models/test_gpt_model.py \
  -k "output_processor or schedule_plan"

2 passed, 14 deselected
```

Full test file:

```text
CUDA_VISIBLE_DEVICES=0 pytest -q tests/unit_tests/models/test_gpt_model.py

13 passed, 1 skipped, 2 failed
```

This matches the clean `main` baseline exactly. The same two custom-process-group tests fail on `main`: both construct an 8-rank `HyperCommGrid` under a local `world_size=1` run, so no PR-introduced regression was found.

Additional checks completed successfully:

- `python3 -m py_compile`
- `git diff --check`
- `isort --check-only`
- `ruff check --no-fix`

The repository-pinned pre-commit and Black toolchain was not available locally, so the official project CI remains required for those checks.

Related to #4590.

Follow-up to #4686.

### CODE REVIEW

**Comment by cspades** [comment]
/ok to test [a42ee1e](https://github.com/NVIDIA/Megatron-LM/pull/5739/commits/a42ee1efe58385775e594ed87386bbc120a5524d)

**Comment by chengcuiping** [comment]
Thanks @cspades for approving the previous `a42ee1e` HEAD for testing.

I rebased this PR onto current `main` at `a54bd8cc`. The new HEAD is `e9da270`. I also consolidated the new contract coverage into the existing `output_processor` test, reducing the patch while retaining coverage for structured callback results, identity-preserving forwarding, keyword-only `context=`, and tied/untied output-weight semantics.

Local validation on one GPU completed with the focused tests passing (`3 passed`). The full test file reported `14 passed, 1 skipped, 2 failed`; a clean `main` run produced the same two custom-process-group failures because they construct an 8-rank grid under `world_size=1`, so no PR regression was found. Compile, diff, isort, and Ruff checks also passed.

A concrete downstream use of the structured callback result is [[verl-project/verl#6933](https://github.com/verl-project/verl/pull/6933)](https://github.com/verl-project/verl/pull/6933).

Could you please run `/ok to test e9da270` for the new HEAD and review when convenient? Thank you.

File: megatron/core/models/gpt/gpt_model.py [hunk 1]
```diff
@@ -546,10 +546,16 @@ def forward(
...
-            output_processor_context (Any, optional): User-defined context object forwarded to
-                `output_processor`.
+            output_processor (Callable, optional): Custom postprocess callback invoked after
+                decoder hidden states are available and before the default logits/loss path.
+                When provided, the callback result becomes the direct `GPTModel.forward` result.
+                The callback receives the context object using the `context` keyword.
+                `output_weight` is populated with the shared embedding/output weight when
+                embeddings and output weights are tied. For untied embeddings, `output_weight`
+                is None and the output projection weight remains available as
+                `output_layer.weight`.
```
Comment (cspades) [inline megatron/core/models/gpt/gpt_model.py]: Doesn't this over-specify what this abstract output process does? Why is this information relevant again?

**Comment by cspades** [comment]
/ok to test [e9da270](https://github.com/NVIDIA/Megatron-LM/pull/5739/commits/e9da270f93f3167e2cab12b584ba9ca316fe1770)

**Comment by cspades** [comment]
Please ping @NVIDIA/mcore-oncall for all future assistance, we rotate.

**Comment by chengcuiping** [comment]
Thanks for approving the new HEAD for testing and for the guidance. I’ll use the rotating MCore on-call team for future CI and review follow-ups.

File: megatron/core/models/gpt/gpt_model.py (hunk 1 above)
Comment (chengcuiping) [reply megatron/core/models/gpt/gpt_model.py]: You're right. Those details describe the current callback invocation mechanics and are useful to downstream processors that reuse the output projection, but they over-specify the abstract `output_processor` contract.

I removed the context keyword and tied/untied `output_weight` details from the public docstring and narrowed the test change accordingly. The PR is now focused on the actual contract mismatch: an output processor may return an arbitrary result, and `GPTModel.forward()` propagates that result directly.

**Comment by chengcuiping** [comment]
@NVIDIA/mcore-oncall, the review feedback has been addressed and the branch was amended to new HEAD `701bab31b`. Could you please run `/ok to test 701bab31b` for the new HEAD and review the PR when convenient? Thank you.

**Comment by janEbert** [comment]
Just to note, oncall is not a required review. @NVIDIA/gpt, @NVIDIA/core-adlr, and @NVIDIA/core-nemo are the required review groups.

**Comment by janEbert** [comment]
/claude fix Also update the type from `Tensor` to `Any` here: https://github.com/chengcuiping/Megatron-LM/blob/701bab31b9631866166c9592c65f2180d068e3ee/megatron/core/models/common/model_chunk_schedule_plan.py#L354

**Comment by chengcuiping** [comment]
Thanks @janEbert for the clarification, and for applying the schedule-plan type fix and re-triggering CI on `2015cfc`. I appreciate the help.

I’ll wait for the checks on the updated HEAD and then follow up with the required review groups.

**Review by jaredcasper — APPROVED** [review]

**Review by ericharper — APPROVED** [review]

**Review by cuichenx — APPROVED** [review]
